### ISSUE TITLE
np.float64(array) returns scalar for any single-element array (unlike np.float32, np.int64, etc.)

### ISSUE THREAD

**Issue by thejohnhoffer** [numpy/numpy#10995]
Hi all, I'm reporting a problem when using `np.float64` as a constructor.

For any `a = np.array([[scalar]])` assuming `type(scalar) in (int, float)`, this raises `AssertionError`:

```python
assert np.float64(a).shape is a.shape
```
The shape after casting to `np.float64` is `()` instead of the expected `(1,1)`. We can see what's going on more clearly when only this last assertion fails:

```python
scalar = 0.0 # example
a = np.array([[scalar]])
assert type(a) is np.ndarray
assert type(np.float64(a)) is np.ndarray #AssertionError
```

The problem is that `np.float64(a) is np.float64(scalar)`. Contrary to expectations, `np.float64(a) is not np.float64([[scalar]])`

This happens with `v1.14.2` as well as `master`. I've tested with python `3.6.4`/`2.7.11` on `ubuntu 16.04`/`Mac OS 10.12.3`.

**Comment by mhvk** [comment]
(Changed the title to reflect more clearly what I think the underlying issue is)

**Comment by thejohnhoffer** [comment]
Thank you for the much more precise title. As you've noted, it's definitely a concern that the behavior of `np.float64` differs so substantially from the other `dtype / ndarray` constructors such as `np.float32` and `np.int64`.

**Comment by eric-wieser** [comment]
The difference is that `issubclass(np.float64, float)` is true, and it seems some of the `float` behavior leaks through.

**Comment by thejohnhoffer** [comment]
@seberg it's great to get an "official" answer on this. It's true we have not found this in the documentation. I feel there is a lot of code that does use these types as casting functions. I even noticed some of the numpy tests rely on them this way.

**Comment by thejohnhoffer** [comment]
The reason I've picked up this bug is that scikit-image uses `np.float64(array)` for their `img_as_float` conversion function, so we noticed this unexpected behavior propogated through that library.

**Comment by thejohnhoffer** [comment]
The [relevant numpy docs page](https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html) doesn't appear to give any mention of how `np.float64` should be called.

**Comment by thejohnhoffer** [comment]
The documentation page linked in my previous comment does give this example, which does suggest `np.int8` can be used as a casting function:

```python
>>> z.astype(float)                 
array([  0.,  1.,  2.])
>>> np.int8(z)
array([0, 1, 2], dtype=int8
```

**Comment by seberg** [comment]
Hmmm, yeah, well I for one would prefer to not suggest the use really, but maybe that is just me. I also just noticed also that `np.cast` dictionary... I do not even find any documentation or usage of it inside numpy on first check. It forces a copy on the input array (which is probably the better default admittedly).
That page could also mention the  `copy=False` option to the `astype` method.

**Comment by thejohnhoffer** [comment]
@eric-wieser do you know where in the source-code this subclassing of `float` by `float64` is defined? There must be somewhere where the `float64(array)` behavior is defined, but somehow `float64` inherits this behavior from `float`:
```
0.0 == float(np.array([[0]]))
```

Of course as @seberg mentioned, the creation of arrays by calling numpy dtypes as constructors is undocumented, so perhaps there is no problem that np.float64 behaves fundamentally differently than np.float32 or np.int64 when called with an ndarray of a single value.

**Comment by eric-wieser** [comment]
This same bug exists for `np.int_` on python 2.7 (where it subclasses `int`)

**Comment by seberg** [comment]
This is fixed in 2.0 (maybe earlier).  Although, IIRC the code is still a bit weird around the edges.
(A simplification may happen pretty naturally, once `__float__()` is fully deprecated.)
